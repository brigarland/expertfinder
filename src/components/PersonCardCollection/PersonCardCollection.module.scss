@import '../../styles/ColorPalette.scss';

.personCardCollectionRoot {
	display: flex;
	flex-wrap: wrap;
}
