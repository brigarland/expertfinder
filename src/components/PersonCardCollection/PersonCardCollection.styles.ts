import { ICheckboxStyles } from 'office-ui-fabric-react'

export interface IPersonCardCollectionStyles {
	customCheckbox: ICheckboxStyles
}

export const jsStyles: IPersonCardCollectionStyles = {
	customCheckbox: {
		root: {
			selectors: {
				'&:hover .ms-Checkbox-checkbox': {
					borderColor: '#323130',
					background: '#fff',
					color: '#6264a7',
				},
			},
		},
		checkbox: {
			background: '#fff',
			color: '#6264a7',
			selectors: {
				'.is-checked &': {
					borderColor: '#323130',
					background: '#fff',
				},
			},
		},
		checkmark: {
			color: '#6264a7',
			selectors: {
				'.is-checked &': {
					fontWeight: 800,
				},
			},
		},
	},
}
