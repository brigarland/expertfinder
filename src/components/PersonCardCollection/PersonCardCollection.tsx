import React, { Fragment, useCallback, useState, useEffect } from 'react'
import { IPerson, MessageContext } from '../../api'
import { PersonCard } from './PersonCard'
import { InfoPanel } from './InfoPanel'
import { ColorPalette } from '../../styles'
import styles from './PersonCardCollection.module.scss'

export interface IPersonCardCollectionProps
	extends React.AllHTMLAttributes<HTMLElement> {
	/**
	 * The main text string value
	 */
	personItems: IPerson[]
	/**
	 * Function to set the selected card ids as an array of strings
	 */
	handleCardSelection: (id: string, checked?: boolean) => void
	/**
	 * Optional Selected Context
	 */
	selectedContext?: MessageContext
}

export const PersonCardCollection: React.FC<IPersonCardCollectionProps> = ({
	personItems,
	handleCardSelection,
	selectedContext,
}) => {
	const rowLength = 4

	const [activeCard, setActiveCard] = useState<number>(-1)
	const [activePerson, setActivePerson] = useState<IPerson>(emptyPerson)
	const [cardClasses, setCardClasses] = useState<string[]>([])
	const [infoPanelVis, setInfoPanelVis] = useState<boolean[]>([])
	const [customCardColor, setCustomCardColor] = useState<string>(
		ColorPalette.brand00,
	)

	// Manages the UI elements to activate and
	// deactivate cards
	const handleCardInteractions = useCallback(
		(cardIndex, clearClasses = false) => {
			// Cards
			const newCardClasses: string[] = [],
				newInfoPanelVis: boolean[] = []
			personItems.forEach((el, i) => {
				newCardClasses.push(
					cardIndex === i && !clearClasses ? 'activeCard' : '',
				)
			})
			setCardClasses(newCardClasses)
			// InfoPanels
			const infoTabsCount = Math.ceil(newCardClasses.length / rowLength),
				infoTabIndex = Math.floor(cardIndex / rowLength)
			console.log('infoTabsCount', infoTabsCount)
			console.log('infoTabIndex', infoTabIndex)
			for (let i = 0; i < infoTabsCount; i++) {
				newInfoPanelVis.push(infoTabIndex === i && !clearClasses ? false : true)
			}

			console.log('newInfoPanelVis', newInfoPanelVis)
			setInfoPanelVis(newInfoPanelVis)
		},
		[personItems, setCardClasses, setInfoPanelVis],
	)

	const handleCardClick = useCallback(
		(i, id) => {
			if (activeCard === i) {
				handleCardInteractions(i, true)
				setActiveCard(-1)
				setActivePerson(emptyPerson)
			} else {
				handleCardInteractions(i)
				const p = personItems.filter(obj => obj.id === id)
				setActiveCard(i)
				setActivePerson(p.length ? p[0] : emptyPerson)
			}
		},
		[
			activeCard,
			setActiveCard,
			setActivePerson,
			personItems,
			handleCardInteractions,
		],
	)

	useEffect(() => {
		switch (selectedContext) {
			case MessageContext.Mentor:
				setCustomCardColor(ColorPalette.mentorPrimary)
				break
			case MessageContext.Expert:
				setCustomCardColor(ColorPalette.expertPrimary)
				break
			case MessageContext.Influencer:
				setCustomCardColor(ColorPalette.influencerPrimary)
				break
			default:
				break
		}
	}, [selectedContext, setCustomCardColor])

	const persontItemsLength = personItems.length

	return (
		<div className={styles.personCardCollectionRoot}>
			{personItems.map((e, i) => {
				const card = (
					<PersonCard
						key={e.id}
						cardIndex={i}
						person={e}
						onClickCard={handleCardClick}
						mainClassNames={cardClasses[i] || ''}
						onChangeCheckbox={handleCardSelection}
						customColor={customCardColor}
					/>
				)
				// console.log(i, infoPanelVis[Math.ceil(i / rowLength) - 1])
				console.log('i % rowLength', i % rowLength)
				if (i % rowLength === 3 || persontItemsLength === i + 1) {
					return (
						<Fragment key={e.id + 1000}>
							{card}
							<InfoPanel
								person={activePerson}
								hidden={infoPanelVis[Math.ceil(i / rowLength) - 1]}
							/>
						</Fragment>
					)
				}
				return card
			})}
		</div>
	)
}

const emptyPerson = {
	id: '',
	email: '',
	function: '',
	organization: '',
	region: '',
	skills: [],
	projects: [],
	topics: [],
	eigenCentrality: 0,
	betweenness: 0,
	pageRank: 0,
	_rid: '',
	_self: '',
	_etag: '',
	_attachments: '',
	_ts: 0,
}
