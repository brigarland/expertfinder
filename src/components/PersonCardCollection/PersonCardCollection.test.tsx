import React from 'react'
import { PersonCardCollection } from './PersonCardCollection'
import renderer from 'react-test-renderer'
it('renders correctly', () => {
	const personItems = [
		{
			id: 'bgarland@bpcs.com',
			email: 'bgarland@bpcs.com',
			function: 'Engineering',
			organization: 'Engineering',
			region: 'West',
			skills: [
				'Management',
				'Pearl',
				'React',
				'Python',
				'Gremlin',
				'Angular',
				'Cosmos',
			],
			projects: [
				'Azure Templates',
				'Bing',
				'Volumetric',
				'Workplace Intelligence',
				'My Analytics',
			],
			topics: [
				'Customer Solutions',
				'C# compilers',
				' Organization Network Analysis',
				' Topic Recommendation',
				'Linear regression',
			],
			eigenCentrality: 0.8,
			betweenness: 0.6,
			pageRank: 0.7,
			_rid: '-jJ8ALmsliQHAAAAAAAAAA==',
			_self: 'dbs/-jJ8AA==/colls/-jJ8ALmsliQ=/docs/-jJ8ALmsliQHAAAAAAAAAA==/',
			_etag: '"0b00dc0a-0000-0700-0000-5da8da920000"',
			_attachments: 'attachments/',
			_ts: 1571347090,
		},
		{
			id: 'aklimov@microsoft.com',
			email: 'aklimov@microsoft.com',
			function: 'Research',
			organization: 'Research',
			region: 'North',
			skills: [
				'Management',
				'Pearl',
				'React',
				'Python',
				'Gremlin',
				'Leadership',
				'Cosmos',
				'Machine learning',
			],
			projects: [
				'Azure Templates',
				'Bing',
				'Volumetric',
				'Teaser',
				'EBC Experience',
			],
			topics: [
				'Machine learning',
				'Artificial Intelligence',
				' Organization Network Analysis',
				' Topic Recommendation',
				'Claims Management',
			],
			eigenCentrality: 0.6,
			betweenness: 0.8,
			pageRank: 0.8,
			_rid: '-jJ8ALmsliQDAAAAAAAAAA==',
			_self: 'dbs/-jJ8AA==/colls/-jJ8ALmsliQ=/docs/-jJ8ALmsliQDAAAAAAAAAA==/',
			_etag: '"0b001e0a-0000-0700-0000-5da8da700000"',
			_attachments: 'attachments/',
			_ts: 1571347056,
		},
	]

	const tree = renderer
		.create(<PersonCardCollection personItems={personItems} />)
		.toJSON()
	expect(tree).toMatchSnapshot()
})
