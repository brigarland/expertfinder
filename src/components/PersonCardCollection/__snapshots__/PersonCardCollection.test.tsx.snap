// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders correctly 1`] = `
<div
  className="personCardCollectionRoot"
>
  <div
    className="personCardRoot"
  >
    <div
      className="personCardMain"
      onClick={[Function]}
      style={
        Object {
          "cursor": "pointer",
        }
      }
    >
      <div
        className="title"
      >
        Engineering
         in 
        Engineering
        <span>
          West
           Region
        </span>
      </div>
      <div
        className="degreeBadgeCnt"
      >
        <div
          className="degreeBadgeRoot"
        >
          <div
            className="valueCnt"
          >
            1
            <sup>
              st
            </sup>
          </div>
        </div>
      </div>
    </div>
    <div
      className="personCardFooter"
    >
      <div
        className="influenceBadgeCnt"
      >
        <div
          className="influenceBadgeRoot"
        >
          <div
            className="IconCnt"
          >
            <img
              alt="Influencer Score Icon"
              height="25"
              src="influence.svg"
              width="25"
            />
          </div>
          <div
            className="valueCnt"
          >
            0.7
          </div>
        </div>
      </div>
      <div
        className="checkBoxCnt"
      >
        <div
          className="ms-Checkbox is-enabled root-49"
        >
          <input
            aria-checked="false"
            checked={false}
            className="input-50"
            id="checkbox-1"
            onBlur={[Function]}
            onChange={[Function]}
            onFocus={[Function]}
            type="checkbox"
          />
          <label
            className="ms-Checkbox-label label-51"
            htmlFor="checkbox-1"
          >
            <div
              className="ms-Checkbox-checkbox checkbox-52"
            >
              <i
                aria-hidden={true}
                className="ms-Checkbox-checkmark checkmark-55"
                data-icon-name="CheckMark"
              />
            </div>
          </label>
        </div>
      </div>
    </div>
  </div>
  <div
    className="personCardRoot"
  >
    <div
      className="personCardMain"
      onClick={[Function]}
      style={
        Object {
          "cursor": "pointer",
        }
      }
    >
      <div
        className="title"
      >
        Research
         in 
        Research
        <span>
          North
           Region
        </span>
      </div>
      <div
        className="degreeBadgeCnt"
      >
        <div
          className="degreeBadgeRoot"
        >
          <div
            className="valueCnt"
          >
            1
            <sup>
              st
            </sup>
          </div>
        </div>
      </div>
    </div>
    <div
      className="personCardFooter"
    >
      <div
        className="influenceBadgeCnt"
      >
        <div
          className="influenceBadgeRoot"
        >
          <div
            className="IconCnt"
          >
            <img
              alt="Influencer Score Icon"
              height="25"
              src="influence.svg"
              width="25"
            />
          </div>
          <div
            className="valueCnt"
          >
            0.8
          </div>
        </div>
      </div>
      <div
        className="checkBoxCnt"
      >
        <div
          className="ms-Checkbox is-enabled root-49"
        >
          <input
            aria-checked="false"
            checked={false}
            className="input-50"
            id="checkbox-2"
            onBlur={[Function]}
            onChange={[Function]}
            onFocus={[Function]}
            type="checkbox"
          />
          <label
            className="ms-Checkbox-label label-51"
            htmlFor="checkbox-2"
          >
            <div
              className="ms-Checkbox-checkbox checkbox-52"
            >
              <i
                aria-hidden={true}
                className="ms-Checkbox-checkmark checkmark-55"
                data-icon-name="CheckMark"
              />
            </div>
          </label>
        </div>
      </div>
    </div>
  </div>
</div>
`;
